src/vs/workbench/contrib/terminalContrib/chat/browser/terminalChatService.ts

```diff
@@ -0,0 +1,159 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Emitter, Event } from '../../../../../base/common/event.js';
+import { Disposable, DisposableMap, IDisposable } from '../../../../../base/common/lifecycle.js';
+import { ILogService } from '../../../../../platform/log/common/log.js';
+import { ITerminalChatService, ITerminalInstance, ITerminalService } from '../../../terminal/browser/terminal.js';
+import { IStorageService, StorageScope, StorageTarget } from '../../../../../platform/storage/common/storage.js';
+import { ILifecycleService } from '../../../../services/lifecycle/common/lifecycle.js';
+import { TerminalCapability } from '../../../../../platform/terminal/common/capabilities/capabilities.js';
+import { PromptInputState } from '../../../../../platform/terminal/common/capabilities/commandDetection/promptInputModel.js';
+
+/**
+ * Used to manage chat tool invocations and the underlying terminal instances they create/use.
+ */
+export class TerminalChatService extends Disposable implements ITerminalChatService {
+	declare _serviceBrand: undefined;
+
+	private static readonly _storageKey = 'terminalChat.toolSessionMappings';
+
+	private readonly _terminalInstancesByToolSessionId = new Map<string, ITerminalInstance>();
+	private readonly _terminalInstanceListenersByToolSessionId = this._register(new DisposableMap<string, IDisposable>());
+	private readonly _onDidRegisterTerminalInstanceForToolSession = new Emitter<ITerminalInstance>();
+	readonly onDidRegisterTerminalInstanceWithToolSession: Event<ITerminalInstance> = this._onDidRegisterTerminalInstanceForToolSession.event;
+
+	/**
+	 * Pending mappings restored from storage that have not yet been matched to a live terminal
+	 * instance (we match by persistentProcessId when it becomes available after reconnection).
+	 * toolSessionId -> persistentProcessId
+	 */
+	private readonly _pendingRestoredMappings = new Map<string, number>();
+
+	constructor(
+		@ILogService private readonly _logService: ILogService,
+		@ITerminalService private readonly _terminalService: ITerminalService,
+		@IStorageService private readonly _storageService: IStorageService,
+		@ILifecycleService private readonly _lifecycleService: ILifecycleService
+	) {
+		super();
+
+		this._restoreFromStorage();
+		this._register(this._lifecycleService.onBeforeShutdown(async e => {
+			let veto = false;
+			// Show all hidden terminals before shutdown so they are restored
+			for (const [toolSessionId, instance] of this._terminalInstancesByToolSessionId) {
+				if (this.terminalIsHidden(toolSessionId) && (instance.capabilities.get(TerminalCapability.CommandDetection)?.promptInputModel.state === PromptInputState.Execute || instance.hasChildProcesses)) {
+					await this._terminalService.showBackgroundTerminal(instance, true, true);
```

Comment:
The await here is a bit of a problem actually. We'll likely get pinged by Ben about slowing shutdown is this lands, especially in its current state.

Maybe a better solution is to add the flag to `IShellLaunchConfig` where we force persistence despite `hideFromUser`? That also comes with the benefit of keeping the terminal hidden.